include/fmt/format-inl.h

```diff
@@ -28,6 +28,14 @@
 
 #include "format.h"
 
+// In non-header-only mode, functions in this file are compiled by format.cc
+// once and exclusively, so we define functions without `inline`;
+// this triggers a clang-tidy warning (misc-definitions-in-headers).
+// If the user is using clang-tidy 14+, we can suppress it.
+#ifndef FMT_HEADER_ONLY
+// NOLINTBEGIN(misc-definitions-in-headers)
+#endif
```

Comment:
I think this should be moved to format.cc and made unconditional.

Reply:
If I move it to format.cc, it doesn't look like it works as intended:
```
$ clang-tidy --checks=misc-definitions-in-headers -p compile_commands.json src/format.cc
15 warnings generated.
Suppressed 15 warnings (15 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
$ clang-tidy --checks=misc-definitions-in-headers -p compile_commands.json -header-filter=.* src/format.cc
15 warnings generated.
/path/to/fmt/include/fmt/format-inl.h:34:15: warning: function 'assert_fail' defined in a header file; function definitions in header files can lead to ODR violations [misc-definitions-in-headers]
FMT_FUNC void assert_fail(const char* file, int line, const char* message) {
              ^
```
which is the same behavior as when the `NOLINT` is not present.  It appears that clang-tidy is hiding the warnings not because they're suppressed, but because it hides warnings in headers by default; and that `NOLINT` is not effective cross-`#include`.

If the `NOLINT` is in format-inl.h, the warnings appear suppressed as intended:
```
$ clang-tidy --checks=misc-definitions-in-headers -p compile_commands.json -header-filter=.* src/format.cc
Suppressed 15 warnings (15 NOLINT).
```

Reply:
Can you run clang-tidy on {fmt} in the header-only mode (with `FMT_HEADER_ONLY` defined)?